src/main/java/com/github/princesslana/eriscasper/rx/websocket/RxWebSocket.java

```diff
@@ -29,31 +31,63 @@ public RxWebSocket(OkHttpClient http) {
             em -> {
               Request rq = new Request.Builder().url(url).build();
 
-              ws = http.newWebSocket(rq, new Listener(em));
+              ws = http.newWebSocket(rq, new Listener(RxWebSocket.this, em));
             })
+        .takeUntil(event -> closed)
         .doOnNext(e -> LOG.trace("Received: {}.", e))
         .doOnError(e -> LOG.warn("Error: {}.", e));
   }
 
+  private void close0() {
+    closed = true;
+  }
+
+  public void close(int code) {
+    close(code, null);
+  }
+
+  public void close(int code, String reason) {
+    close0();
+    ws.close(code, reason);
+  }
+
+  public Completable closeDueToInvalidSession() {
+    return Completable.fromAction(
+        () -> {
+          close(1002, "Invalid session.");
+        });
+  }
+
   public Completable send(String text) {
     return Completable.fromAction(() -> ws.send(text))
         .doOnComplete(() -> LOG.trace("Sent: {}.", text));
   }
 
   private static class Listener extends WebSocketListener {
     private final ObservableEmitter<RxWebSocketEvent> em;
+    private final RxWebSocket socket;
 
-    public Listener(ObservableEmitter<RxWebSocketEvent> em) {
+    private Listener(RxWebSocket socket, ObservableEmitter<RxWebSocketEvent> em) {
+      this.socket = socket;
       this.em = em;
     }
 
     @Override
     public void onClosed(WebSocket ws, int code, String reason) {
       em.onNext(ClosedTuple.of(ws, code, reason));
+      socket.close0();
     }
 
     @Override
     public void onClosing(WebSocket ws, int code, String reason) {
+      if (!socket.closed) {
+        if (code == 4004) {
+          socket.close(1002, "Invalid token.");
+          new ErisCasperFatalException("Failed to authenticate with discord servers.")
+              .fillInStackTrace()
+              .printStackTrace();
```

Comment:
Don't print to stdout.

Why `fillInStackTrace`? Did you mean to `throw` this exception?

Reply:
throwing it leaves a long very ambiguous looking error, seems to leave a much cleaner looking error if we directly print it, more useful to the end user imo, but I can change it.

Reply:
If we can't continue, then throwing an exception is the best way to indicate this

Reply:
Alright, fair enough, I'll throw it rather than printing it.